### ISSUE TITLE
do_predict?

### ISSUE THREAD

**Issue by acmilannesta** [kamalkraj/ALBERT-TF2.0#3]
Hi,

Can the script do predict? I may miss it, but I didn't see a "do_pred" flag.

**Comment by kamalkraj** [comment]
Hi @acmilannesta,
do_predict added.

**Comment by acmilannesta** [comment]
Great, thx so much!

One more issue for "create_finetuning_data.py", I got the following error when I run in colab:
```
    !python /content/ALBERT-TF2.0/create_finetuning_data.py \
        --input_data_dir=/content/input \
        --spm_model_file=/content/drive/My\ Drive/albert_model/large/vocab/30k-clean.model \
        --train_data_output_path=/content/input/train.tf_record \
        --eval_data_output_path=/content/input/eval.tf_record \
        --meta_data_file_path=/content/input/meta_data \
        --fine_tuning_task_type=classification \
        --max_seq_length=142 \
        --classification_task_name=COLA

Traceback (most recent call last):
  File "/content/ALBERT-TF2.0/create_finetuning_data.py", line 149, in <module>
    app.run(main)
  File "/usr/local/lib/python3.6/dist-packages/absl/app.py", line 299, in run
    _run_main(main, args)
  File "/usr/local/lib/python3.6/dist-packages/absl/app.py", line 250, in _run_main
    sys.exit(main(argv))
  File "/content/ALBERT-TF2.0/create_finetuning_data.py", line 137, in main
    input_meta_data = generate_classifier_dataset()
  File "/content/ALBERT-TF2.0/create_finetuning_data.py", line 122, in generate_classifier_dataset
    do_lower_case=FLAGS.do_lower_case)
  File "/content/ALBERT-TF2.0/classifier_data_lib.py", line 832, in generate_tf_record_from_data_file
    train_input_data_examples = processor.get_train_examples(data_dir)
  File "/content/ALBERT-TF2.0/classifier_data_lib.py", line 269, in get_train_examples
    self._read_tsv(os.path.join(data_dir, "CoLA", "train.tsv")), "train")
  File "/content/ALBERT-TF2.0/classifier_data_lib.py", line 113, in _read_tsv
    for line in reader:
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/lib/io/file_io.py", line 220, in __next__
    return self.next()
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/lib/io/file_io.py", line 214, in next
    retval = self.readline()
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/lib/io/file_io.py", line 178, in readline
    self._preread_check()
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/lib/io/file_io.py", line 84, in _preread_check
    compat.as_bytes(self.__name), 1024 * 512)
tensorflow.python.framework.errors_impl.NotFoundError: /content/input/CoLA/train.tsv; No such file or directory
```
It seems to ask for a CoLA directory, while classification_name only accepts "COLA".

**Comment by acmilannesta** [comment]
Great, thx so much!

But do you also need to add something like prediction data output in "create_finetuning_data.py"?

**Comment by kamalkraj** [comment]
@acmilannesta 
for running predict, additional flags
```bash
--do_predict --predict_data_path=${OUTPUT_DIR}/${TASK_NAME}_predict.tf_record --input_data_dir=${GLUE_DIR}/
```

Full cmd
```bash
python run_classifer.py --train_data_path=${OUTPUT_DIR}/${TASK_NAME}_train.tf_record --eval_data_path=${OUTPUT_DIR}/${TASK_NAME}_eval.tf_record --input_meta_data_path=${OUTPUT_DIR}/${TASK_NAME}_meta_data --albert_config_file=${ALBERT_DIR}/config.json --task_name=${TASK_NAME} --spm_model_file=${ALBERT_DIR}/vocab/30k-clean.model --output_dir=${MODEL_DIR} --init_checkpoint=${ALBERT_DIR}/tf2_model.h5 --do_train --do_eval --train_batch_size=128 --learning_rate=1e-5 --custom_training_loop --do_predict --predict_data_path=${OUTPUT_DIR}/${TASK_NAME}_predict.tf_record --input_data_dir=${GLUE_DIR}/
```

**Comment by acmilannesta** [comment]
Epoch 1/3
7698/7698 [==============================] - 9878s 1s/step - loss: 0.8664 - accuracy: 0.7629 - val_loss: 0.0000e+00 - val_accuracy: 0.0000e+00
Epoch 2/3
2445/1925 [======================================] - 2886s 1s/step - loss: 0.5667 - accuracy: 0.8351

Got it. Now I got into another issue. It seems the 2nd training epoch will use my validation set. My training set has 7698 batches, and my validation set has 1925 batches.

**Comment by kamalkraj** [comment]
Second epoch won't use validation dataset as training set, even though it seems like that.
Model.fit api has known issues when tf.data is given as input , steps calculation issues.
Only after first epoch model will know the complete data size . Ignore it.

You can also train using `--custom_training_loop`

**Comment by acmilannesta** [comment]
Thank you!
I've just run my model on AWS g4dn.xlarge (Tesla T4 GPU) VM. However, it almost took 3 times longer than running bert-base model. Here is my code:

```
    os.system(
    '/home/ubuntu/py3env/bin/python3.6 ALBERT-TF2.0/run_classifer.py \
        --train_data_path=input/COLA_train.tf_record \
        --eval_data_path=input/COLA_eval.tf_record \
        --predict_data_path= input/COLA_predict.tf_record \
        --input_data_dir=input/CoLA \
        --input_meta_data_path=input/COLA_meta_data \
        --albert_config_file=base_2/config.json \
        --task_name=CoLA \
        --spm_model_file=base_2/vocab/30k-clean.model \
        --output_dir=output \
        --init_checkpoint=base_2/tf2_model.h5 \
        --do_train \
        --do_eval \
        --do_predict \
        --train_batch_size=16 \
        --eval_batch_size=16 \
        --learning_rate=5e-5 \
        --max_seq_length=142 \
        --num_train_epochs=3 \
        --custom_training_loop'
    )
```

**Comment by kamalkraj** [comment]
@acmilannesta 
How much time for bert and Albert ?

**Comment by acmilannesta** [comment]
About 170-175 ms/step on bert
480-490 ms/step on albert

Kamal Raj <notifications@github.com> 于2019年11月11日周一 上午8:10写道：

**Comment by kamalkraj** [comment]
Which implementation of bert are you using ?
Is batch size same in bert and albert training ?

**Comment by acmilannesta** [comment]
I used a package called keras_bert (https://github.com/CyberZHG/keras-bert), whcih basically load ckpt into keras model.

Yes, the batch size is same = 16.

**Comment by kamalkraj** [comment]
TensorFlow version for keras_bert ?

**Comment by acmilannesta** [comment]
Yes.

**Comment by kamalkraj** [comment]
Tensorflow 1.X or 2.0 ?

**Comment by acmilannesta** [comment]
I think both work. But I'm running on 1.x

**Comment by kamalkraj** [comment]
Tensorflow 2 is slower than 1.x
https://github.com/tensorflow/tensorflow/issues/33487

**Comment by kamalkraj** [comment]
https://github.com/tensorflow/models/tree/master/official/nlp/bert
TF 2.0 implementation of BERT

**Comment by acmilannesta** [comment]
I see. Let me try this
https://github.com/google-research/google-research/tree/master/albert
It seems work on tensorflow 1.x

**Comment by kamalkraj** [comment]
The above one runs on TensorFlow 1.15

**Comment by acmilannesta** [comment]
I just tried. About 2.4 steps/second (416 ms/step).

**Comment by acmilannesta** [comment]
https://github.com/tensorflow/tensorflow/issues/33487#issuecomment-543652394

Based on this comment, would you suggest turning off eager in tf2.0 to run your codes?

**Comment by kamalkraj** [comment]
@acmilannesta 
Try and let me know the results

**Comment by acmilannesta** [comment]
Unfortunately, it doesn't allow for disabling eager in tf2.0.


Kamal Raj <notifications@github.com> 于2019年11月11日周一 下午1:01写道：
